docs/proposals/015-bindless-t.md

```diff
@@ -0,0 +1,160 @@
+
+With the recent advancement in hardware capabilities, a lot of modern graphics systems are adopting a "bindless" parameter passing idiom, where all resource handles are passed to the shader in a single global array, and all remaining references to texture, buffers or sampler states are represented as a single integer index into the array. This allows the shader code to workaround the restrictions around the opaque handle types.
+
+Direct3D Shader Model 6.6 introduces the "Dynamic Resources" capability, which further simplifies the way to write bindless shader code by removing the need to even declare the global array.
+
+We believe that graphics developers will greatly benefit from a system defined programming model for the bindless parameter passing idom that is versatile and cross-platform, which will provide a consistent interface so that different shader libraries using the bindless pattern can interop with each other without barriers.
+
+## Proposed Approach
+
+We introduce a `Bindless<T>` type that is defined as:
+```
+struct Bindless<T> : IComparable<Bindless<T>>
+    where T : IOpaqueHandle
+{
+    __init(uint64_t value);
+}
+```
+Where `IOpaqueHandle` is an interface that is implemented by all texture, buffer and sampler state types:
+
+```slang
+enum ResourceKind
+{
+    Unknown, Texture, ConstantBuffer, StorageBuffer, Sampler
+}
+interface IOpaqueHandle
+{
+    static ResourceKind getResourceKind();
+}
+```
+
+### Basic Usage
+
+`Bindless<T>` should provide the following features:
+
+- Construction/explicit cast from a 64-bit integer.
+- Explicit cast to a 64-bit integer.
+- Equality comparison.
+- Implicit dereference to `T`.
+- Implicit conversion to `T`.
+
+For example:
+
+```slang
+uniform Bindless<Texture2D> texture;
+uniform Bindless<SamplerState> sampler;
+
+void test()
+{
+    // Explicit cast from bindless handle to uint64_t value.
+    let idx = (uint64_t)texture;
+
+    // Constructing bindless handle from uint64_t value.
+    let t = Bindless<Texture2D>(idx);
+
+    // Comparison.
+    ASSERT(t == texture);
+
+    // OK, `t` is first implicitly dereferenced to producee `Texture2D`, and
+    // then `Texture2D::Sample` is called.
+    // The `sampler` argument is implicitly converted from `Bindless<SamplerState>`
+    // to `SamplerState`.
+    t.Sample(sampler, float2(0,0));
+}
+```
+
+A `Bindless<T>` type is a concrete type whose size is always 8 bytes and is internally a 64-bit integer.
+This means that you can use a `Bindless<T>` type in any context where an ordinary data type, e.g. `int` type
+is allowed, such as in buffer elements.
+
+On targets where resource handles are already concrete and sized types, `Bindless<T>` simply translates to just `T`.
```

Comment:
This statement might need some clarification on what happens if the opaque resource types have a size/alignment other than 8 bytes. Ideally, I would expect that `Bindless<T>` is always stored with 8-byte size/alignment, even on targets where `T` might have smaller size/alignment.

That still leaves the open question of what happens if `T` has a native size or alignment that is greater than 8 bytes...

Reply:
Done.